4단계 - 자동차 경주(우승자)

Eundms opened the pull request:
- 자동차 경주게임 Step4 제출합니다

Eundms committed:
3단계 리뷰 반영

- Static vs 인스턴스 메서드 트레이드오프

  Static 메서드의 장점

  - 불필요한 객체 생성 방지
  - 메모리 효율적
  - 유틸리티성 기능에 적합

  Static 메서드의 단점

  1. 테스트가 어려움: 모킹(mocking)이 어렵고, System.out을 검증하기 힘듦
  2. 확장성 저하: 나중에 상태가 필요해지면 모든 호출부를 수정해야 함
  3. 다형성 활용 불가: 인터페이스 구현이나 상속을 통한 교체가 불가능
  4. OOP 원칙 위배: 객체지향 설계에서 벗어남

  인스턴스 메서드의 장점

  1. 테스트 용이성: View 인터페이스를 만들고 테스트용 구현체로 교체 가능
  2. 확장성: 나중에 로그 파일 출력, GUI, Web 등으로 교체 가능
  3. 의존성 주입: DI 컨테이너와 함께 사용하기 좋음

Eundms committed:
요구사항 : Car 5자 이하의 이름, 쉼표로 입력받음

Eundms committed:
요구사항 : 우승자 발표

Eundms committed:
리팩토링 : get으로 꺼낸 값들(tryCnt, position) 객체 안으로 넣고 객체 안에서 작업하기

- 매 라운드 출력이 필요하니 이것도 애매함

javajigi commented on src/main/java/racing/domain/Car.java:
```diff
@@ -1,28 +1,66 @@
+ */
+public class Car implements Comparable<Car> {
+  private static final int MOVE_BORDER = 4;
+  private CarName name;
```
```suggestion
  private final CarName name;
```

javajigi commented on src/main/java/racing/domain/Car.java:
```diff
@@ -1,28 +1,66 @@
+public class Car implements Comparable<Car> {
+  private static final int MOVE_BORDER = 4;
+  private CarName name;
   private int position;
```
자동차 이름에 대해 문자열 포장 원칙을 적용했으니 `int position`에 대해 원시 값을 포장한다는 원칙을 적용해 보면 어떨까요?

javajigi commented on src/main/java/racing/domain/Car.java:
```diff
@@ -1,28 +1,66 @@
+  }
+
+  public Car(CarName name){
+    this(name, 0);
```
부생성자 -> 주생성자 호출 👍

javajigi commented on src/main/java/racing/domain/Car.java:
```diff
@@ -1,28 +1,66 @@
+
+  public Car(CarName name, int position) {
+    this.name = name;
     this.position = position;
```
position 값이 0 이상이어야 한다는 유효성 체크를 해야하지 않을까?

javajigi commented on src/main/java/racing/domain/Car.java:
```diff
@@ -1,28 +1,66 @@
-  private boolean isMove(int value) {
-    return value >= 4;
+  public boolean hasSamePosition(Car topCar) {
+    return this.position == topCar.getPosition();
```
```suggestion
    return this.position == topCar.position;
```
이렇게 구현도 가능함

javajigi commented on src/main/java/racing/domain/CarName.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.Objects;
+import racing.exception.RacingException;
+
+public class CarName {
```
👍

javajigi commented on src/main/java/racing/domain/ErrorMessage.java:
```diff
@@ -0,0 +1,16 @@
+package racing.domain;
+
+public enum ErrorMessage {
```
이름과 관련된 메시지라면 CarName에 위치해도 되지 않을까?
굳이 둘을 분리할 필요가 있을까?

javajigi commented on src/main/java/racing/domain/Participants.java:
```diff
@@ -0,0 +1,48 @@
+  public Participants(List<CarName> carNames){
+    this.cars = initCars(carNames);
+  }
+
```
```suggestion
     Participants(List<Car> cars)
```
주생성자는 기본으로 구현할 것을 추천
객체의 필드 값은 주생성자만 초기화할 수 있도록 구현하는 습관을 들인다.

javajigi commented on src/main/java/racing/domain/Participants.java:
```diff
@@ -0,0 +1,48 @@
+    return cars;
+  }
+
+  public List<Car> getWinners(){
```
추상화 레벨을 맞춰 구현해 보면 어떨까?
findTopCar() 추상화 레벨은 1, 나머지 코드는 0이다.
메서드를 분리해 추상화 레벨을 모두 1로 맞춰 구현해 보면 어떨까?

javajigi commented on src/main/java/racing/domain/MovementStrategy.java:
```diff
@@ -3,9 +3,11 @@
   private final int MIN_VALUE;
   private final int MAX_VALUE;
-  public MovementStrategy(int minMove, int maxMove){
+
```
```suggestion
  public MovementStrategy() {
      this(0, 9);
  }
```
기본 생성자를 추가하면 외부에서 이 객체를 사용하는 개발자 입장에서 더 편리하게 사용할 수 있지 않을까?

javajigi commented on src/main/java/racing/domain/Game.java:
```diff
@@ -1,38 +1,36 @@
+
+  private final MovementStrategy strategy;
+  private final Participants participants;
   private final int tryCnt;
```
int tryCnt에 대해 원시 값을 포장한다는 원칙을 적용해 보면 어떨까요?

javajigi reviewed (changes requested):
드디어 4단계 미션 구현했군요.
3단계에서 객체 설계가 깔끔해 큰 변화없이 4단계 구현했네요. 💯 
일정 수준의 객체 설계 역량을 갖추어 있어 '규칙 3: 모든 원시값과 문자열을 포장한다.' 원칙을 적용해 보면 좋겠습니다.
이 원칙을 지키면서 가능하면 포장한 객체의 값을 꺼내지 말고, 객체에 메시지를 보내도록 사고해 볼 것을 추천합니다.

Eundms committed:
Position 추상화, 추상화 레벨 맞추기, 기본 생성자 추가하여 편리하게 초기화할 수 있도록함, Car에 생성자 추가 , Participants 생성자와 정적팩터리메서드, TryCnt 추상화

-> TODO: getter 없애기, 도메인 테스트 보강 및 이동

Eundms commented:
아직 객체에 메시지를 보내는 부분에 있어 어려움이 느껴지네요.  ㅠ

javajigi commented on src/main/java/racing/domain/Car.java:
```diff
@@ -1,28 +1,65 @@
+    if (value < MOVE_BORDER) {
+      return;
+    }
+    this.position.move(1);
```
```suggestion
    if (value > MOVE_BORDER) {
      this.position.move(1);
    }
```

javajigi replied:
```suggestion
    this.position.increase();
```
외부에서 값을 전달하기보다 1 증가하는 부분을 위와 같이 추상화하도록 의식적인 연습을 해본다.

javajigi commented on src/main/java/racing/domain/CarName.java:
```diff
@@ -0,0 +1,57 @@
+    return this.name;
+  }
+
+  enum CarNameMessage implements ErrorMessage {
```
ErrorMessage을 구현하는 여러 구현 클래스가 있다.
ErrorMessage을 구현하는 클래스에 중복 코드가 많다.
이 중복 코드를 제거하는 방법은 없을까?

javajigi commented on src/main/java/racing/domain/Participants.java:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  public static Participants from(List<String> names) {
+    return new Participants(names.stream().map(Car::new).collect(Collectors.toList()));
```
```suggestion
    return new Participants(names.stream().map(Car::new).toList());
```

javajigi commented on src/main/java/racing/domain/Participants.java:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  private Position findMaxPosition() {
+    return Collections.max(cars, Comparator.comparing(Car::getPosition)).getPosition();
```
객체의 값을 꺼내지 말고 메시지를 보내 구현해 보면 어떨까?

Eundms replied:
객체에 equals(), toString()을 오버라이드하니 getter를 사용할 일이 줄어들었습니다.

javajigi commented on src/main/java/racing/domain/RacingGame.java:
```diff
@@ -0,0 +1,23 @@
+
+  public void play() {
+    ResultView.printResultMessage();
+    for (int i = 0; i < tryCnt.getCount(); i++) {
```
```suggestion
    for (tryCnt.racing()) {
```
위와 같이 메시지를 보내보면 어떨까?
즉, play를 할 때마다 tryCnt의 값을 1씩 감소시켜 0이면 race가 끝...
객체 지향적인 사고 연습해 보면 어떨까?

javajigi commented on src/main/java/racing/GameApplication.java:
```diff
@@ -1,27 +1,24 @@
-    }
-
-    input.close();
+    MovementStrategy strategy = new MovementStrategy(0, 9);
```
```suggestion
    MovementStrategy strategy = new MovementStrategy();
```
빈 생성자 추가했으니 활용해 구현하면 어떨까?

javajigi commented on src/test/java/racing/domain/CarTest.java:
```diff
@@ -1,29 +1,47 @@
+  @CsvSource({"4,1", "9,1", "0,0", "3,0"})
+  void carGoIfMoreThan4(int randomValue, int expected) {
+    car.move(randomValue);
+    assertThat(car.getPosition().getPosition()).isEqualTo(expected);
```
```suggestion
    assertThat(car).isEqualTo(new Car(NAME, expected));
```
테스트 코드의 검증 또한 값을 꺼내기 보다 객체 자체로 비교하도록 의식적인 연습을 해보면 어떨까?

Eundms replied:
말씀주신대로 객체끼리 비교하니 get을 사용해야 할 일이 줄어들었습니다.

javajigi commented on src/test/java/racing/domain/GameTest.java:
```diff
@@ -2,34 +2,23 @@
-    assertThat(game.getTryCnt()).isEqualTo(TRY_CNT);
+  void init() {
+    List<String> carNames = List.of("pobi", "crong", "honux");
+    game = new Game(new MovementStrategy(0, 9), Participants.from(carNames));
```
```suggestion
    game = new Game(new MovementStrategy(), Participants.from(carNames));
```

Eundms replied:
```
  public boolean racing() {
    if(cnt == 0){
      return false;
    }
    cnt -= 1;
    return true;
  }
```
처음에 이런 코드를 생각했지만 상태를 변경하면서 동시에 값을 반환하는 것이 마음에 걸렸습니다.
따라서 canRace()와 doRace() 2가지로 메서드를 나눴습니다.

javajigi commented on src/test/java/racing/domain/GameTest.java:
```diff
@@ -2,34 +2,23 @@
+  void gameReturnsWinner(){
+    List<String> winners = game.getWinners();
+    assertThat(winners).isNotEmpty();
   }
```
Participants에 대한 단위 테스트를 추가한 이 미션의 핵심 로직이라 할 수 있는 우승자를 구하는 단위 테스트를 추가해 보면 어떨까?

Eundms replied:
단위 테스트를 작성할 때, 어떤 상황에서 문제가 발생할 수 있을지 떠올리는 능력이 부족하다는 걸 자주 느낍니다.
정상 케이스는 쉽게 생각하지만, 경계값이나 예외 상황, 그리고 미래 변경에 의해 깨질 수 있는 케이스는 잘 잡아내지 못합니다.
경험 많은 분들이 어떤 방식으로 테스트 시나리오를 떠올리는지 궁금합니다.

javajigi reviewed (changes requested):
피드백 빠르게 반영했네요. 👍 

아직까지 객체를 분리하기는 하는데 객체에 메시지를 보내는 부분에서 어려움을 겪고 있는 것 같아요.
객체에 메시지를 보내는 부분에 대해 피드백 남겼으니 반영해 보세요.
Participants에 대한 단위 테스트도 추가해 우승자 로직에 대한 단위 테스트도 추가해 보면 좋겠네요.
피드백 반영할 때 무조건 빠르게 반영하는 것이 좋은 것이 아니라 피드백 반영을 통해 다른 코드를 어떻게 리팩터링할 수 있는지도 고민해 보면서 리팩터링해보면 좋겠습니다.

Eundms committed:
getter 없애기, 도메인 테스트 보강

- 생성자 오버라이드, equalsTo, hashCode, toString 적극 사용
- jacoco 테스트 커버리지 확인하며 getter 없앰

Eundms committed:
RacingGame 삭제 -> RacingGameApplication이 컨트롤러 역할 대신함

javajigi commented on src/main/java/racing/domain/Car.java:
```diff
@@ -26,14 +29,17 @@ public Car(CarName name, Position position) {
-    if (value < MOVE_BORDER) {
-      return;
+    if (value > MOVE_BORDER) {
+      this.position.increase();
```
👍

javajigi commented on src/main/java/racing/domain/RoundCount.java:
```diff
@@ -1,34 +1,52 @@
 
-  public int getCount() {
-    return cnt;
+  public void doRace() {
```
👍

javajigi commented on src/main/java/racing/domain/RoundCount.java:
```diff
@@ -1,34 +1,52 @@
-    INVALID_ROUND("1번 이상 수행되어야 합니다");
-
-    private String message;
+  public boolean canRace() {
```
👍

javajigi commented on src/main/java/racing/exception/ErrorMessage.java:
```diff
@@ -1,6 +1,11 @@
-public interface ErrorMessage {
-
-  String getMessage();
+public abstract class ErrorMessage {
```
👍 
특히 추상 클래스로 구현해 직접 생성하지 못하도록 구현한 점 👍

javajigi commented on src/test/java/racing/domain/ParticipantsTest.java:
```diff
@@ -0,0 +1,65 @@
+
+  @Test
+  @DisplayName("가장 멀리 이동한 차만 우승자")
+  void onlyMaxPositionCarIsWinner() {
```
👍

javajigi reviewed (approved):
4단계 피드백 반영하는데 여러 어려움이 있었을 것 같은데요.
포기하지 않고 끝까지 반영 👍 

특히 이번 피드백에 우승자 구하는 부분에 대한 단위 테스트 추가 👍 
4단계에서 연습할 부분은 충분히 연습한 것 같아 merge해요.
5단계 연습 방법 확인하고 추가로 연습해 볼 부분을 찾아 연습해 볼 것을 추천합니다.

Q1. 정상 케이스는 쉽게 생각하지만, 경계값이나 예외 상황, 그리고 미래 변경에 의해 깨질 수 있는 케이스는 잘 잡아내지 못합니다.
경험 많은 분들이 어떤 방식으로 테스트 시나리오를 떠올리는지 궁금합니다.
A1. 이 또한 경험이 쌓이다보면 좋아지는 것도 있는데요.
이 역량을 빠르게 키우는 좋은 방법은 소스 코드를 구현하기 전에 요구사항을 철저히 분석해 기능 목록을 만드는 것이라 생각해요.
아무래도 소스 코드 구현은 하지 않고 기능 목록만을 작성하기 때문에 예외 사항을 찾는데 도움이 된다 생각해요.
특히나 요즘은 AI를 활용해 기능 목록을 작성해 달라고 제안하고, AI가 작성한 기능 목록 중 의미있다 생각하는 부분, 빠트린 부분을 찾기 위한 목적으로 활용해도 좋을 것 같아요.
